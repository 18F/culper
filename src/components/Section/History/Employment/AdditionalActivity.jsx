import React from 'react'
import { i18n } from '../../../../config'
import { ValidationElement, Show, BranchCollection, DateRange, Text, Field } from '../../../Form'

export default class AdditionalActivity extends ValidationElement {
  constructor (props) {
    super(props)
    this.updateBranch = this.updateBranch.bind(this)
  }

  update (queue) {
    this.props.onUpdate({
      List: this.props.List,
      ...queue
    })
  }

  updateBranch (values) {
    this.props.onUpdate({ List: values })
  }

  render () {
    const klass = `activity ${this.props.className || ''}`.trim()

    return (
      <div className={klass}>
        <BranchCollection label={i18n.t('history.employment.default.additionalActivity.label')}
                          labelSize="h4"
                          appendLabel={i18n.t('history.employment.default.additionalActivity.collection.append')}
                          appendSize="h4"
                          {...(this.props.List || {})}
                          onUpdate={this.updateBranch}
                          onError={this.props.onError}
                          scrollIntoView={this.props.scrollIntoView}>
          <Field title={i18n.t('history.employment.default.additionalActivity.heading.position')}
                 titleSize="h4"
                 adjustFor="labels"
                 scrollIntoView={this.props.scrollIntoView}>
            <Text name="Position"
                  className="text"
                  label={i18n.t('history.employment.default.additionalActivity.position.label')}
                  bind={true}
                  required={this.props.required}
                  />
          </Field>

          <Field title={i18n.t('history.employment.default.additionalActivity.heading.supervisor')}
                 titleSize="h4"
                 adjustFor="labels"
                 scrollIntoView={this.props.scrollIntoView}>
            <Text name="Supervisor"
                  className="text"
                  label={i18n.t('history.employment.default.additionalActivity.supervisor.label')}
                  bind={true}
                  required={this.props.required}
                  />
          </Field>

          <Field title={i18n.t('history.employment.default.additionalActivity.heading.datesEmployed')}
                 titleSize="h4"
                 help="history.employment.default.additionalActivity.datesEmployed.help"
                 adjustFor="daterange"
                 shrink={true}
                 scrollIntoView={this.props.scrollIntoView}>
            <DateRange name="DatesEmployed"
                       bind={true}
                       applicantBirthdate={this.props.applicantBirthdate}
                       required={this.props.required}
                       />
          </Field>
        </BranchCollection>
      </div>
    )
  }
}

AdditionalActivity.defaultProps = {
  List: {},
  applicantBirthdate: {},
  onUpdate: (queue) => {},
  onError: (value, arr) => { return arr }
}
