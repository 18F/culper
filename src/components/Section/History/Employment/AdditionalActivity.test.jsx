import React from 'react'
import { shallow, mount } from 'enzyme'
import configureMockStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import AdditionalActivity from './AdditionalActivity'

describe('The employment additional activity component', () => {
  it('renders without errors', () => {
    const component = shallow(<AdditionalActivity />)

    expect(component.exists()).toBe(true)
    expect(component).toMatchSnapshot()
  })

  const mockStore = configureMockStore()

  const mountComponentWithStore = (props = {}, defaultState = {}) => {
    const store = mockStore({
      application: {},
      ...defaultState,
    })

    return mount(
      <Provider store={store}>
        <AdditionalActivity {...props} />
      </Provider>
    )
  }

  it('renders the default state', () => {
    const component = mountComponentWithStore()

    expect(component.find('.branch .yes').length).toBe(1)
    expect(component.find('.branch .no').length).toBe(1)
  })

  it('implements an onUpdate handler', () => {
    const onUpdate = jest.fn()
    const expected = {
      onUpdate,
      items: [
        {
          Item: {
            Has: { value: 'Yes' },
            Position: {
              name: 'Position',
              value: 'Dev',
            },
          },
        },
      ],
    }

    const component = mountComponentWithStore(expected)
    component
      .find('.branch .yes input')
      .at(0)
      .simulate('change')
    component.find({ type: 'text', name: 'Position' }).simulate('change')
    expect(onUpdate.mock.calls.length).toBe(2)
  })
})
