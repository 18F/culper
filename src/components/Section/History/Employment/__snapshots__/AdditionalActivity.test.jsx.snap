// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The employment additional activity component renders without errors 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="activity"
  >
    <BranchCollection
      List={Object {}}
      appendContent={null}
      appendLabel="Add additional periods"
      appendSize="h4"
      branchClassName=""
      branchName="branchcollection"
      content={null}
      employmentFromDate={
        Object {
          "day": "15",
          "month": "4",
          "year": "2019",
        }
      }
      help=""
      items={Array []}
      label="Do you have any additional periods of activity to add?"
      labelSize="h4"
      onError={[Function]}
      onUpdate={[Function]}
      removable={true}
      scrollToBottom=""
      valueKey="Has"
    >
      <AccordionItem
        bind={true}
        defaultState={true}
        dispatch={[Function]}
        name="Item"
        onError={[Function]}
        onUpdate={[Function]}
        required={false}
        scrollIntoView={false}
      >
        <Field
          adjustFor="labels"
          className=""
          comments={false}
          commentsActive={false}
          commentsAdd="comments.add"
          commentsName="Comments"
          commentsRemove="comments.remove"
          commentsRequired={false}
          commentsValue={Object {}}
          errorPrefix=""
          errors={Array []}
          filterErrors={[Function]}
          help=""
          helpActive={false}
          optional={false}
          optionalText=""
          scrollIntoView={true}
          shrink={false}
          title="Provide position title"
          titleSize="h4"
          validate={true}
        >
          <Text
            bind={true}
            className="text"
            label="Position"
            name="Position"
            onError={[Function]}
            onUpdate={[Function]}
            prefilter={[Function]}
            prefix=""
            required={false}
            status={true}
            value=""
          />
        </Field>
        <Field
          adjustFor="labels"
          className=""
          comments={false}
          commentsActive={false}
          commentsAdd="comments.add"
          commentsName="Comments"
          commentsRemove="comments.remove"
          commentsRequired={false}
          commentsValue={Object {}}
          errorPrefix=""
          errors={Array []}
          filterErrors={[Function]}
          help=""
          helpActive={false}
          optional={false}
          optionalText=""
          scrollIntoView={true}
          shrink={false}
          title="Provide supervisor"
          titleSize="h4"
          validate={true}
        >
          <Text
            bind={true}
            className="text"
            label="Supervisor"
            name="Supervisor"
            onError={[Function]}
            onUpdate={[Function]}
            prefilter={[Function]}
            prefix=""
            required={false}
            status={true}
            value=""
          />
        </Field>
        <Field
          adjustFor="daterange"
          className=""
          comments={false}
          commentsActive={false}
          commentsAdd="comments.add"
          commentsName="Comments"
          commentsRemove="comments.remove"
          commentsRequired={false}
          commentsValue={Object {}}
          errorPrefix=""
          errors={Array []}
          filterErrors={[Function]}
          help="history.employment.default.additionalActivity.datesEmployed.help"
          helpActive={false}
          optional={false}
          optionalText=""
          scrollIntoView={true}
          shrink={true}
          title="Provide dates of employment"
          titleSize="h4"
          validate={true}
        >
          <DateRange
            allowPresent={false}
            bind={true}
            dateRangePrefix="additionalActivity"
            disabled={false}
            from={
              Object {
                "touched": false,
              }
            }
            maxDate={
              Object {
                "day": "15",
                "month": "4",
                "year": "2019",
              }
            }
            maxDateEqualTo={true}
            minDate={null}
            minDateEqualTo={true}
            name="DatesEmployed"
            onError={[Function]}
            prefix=""
            present={false}
            relationship=""
            to={
              Object {
                "touched": false,
              }
            }
          />
        </Field>
      </AccordionItem>
    </BranchCollection>
  </div>,
  "nodes": Array [
    <div
      className="activity"
    >
      <BranchCollection
        List={Object {}}
        appendContent={null}
        appendLabel="Add additional periods"
        appendSize="h4"
        branchClassName=""
        branchName="branchcollection"
        content={null}
        employmentFromDate={
          Object {
            "day": "15",
            "month": "4",
            "year": "2019",
          }
        }
        help=""
        items={Array []}
        label="Do you have any additional periods of activity to add?"
        labelSize="h4"
        onError={[Function]}
        onUpdate={[Function]}
        removable={true}
        scrollToBottom=""
        valueKey="Has"
      >
        <AccordionItem
          bind={true}
          defaultState={true}
          dispatch={[Function]}
          name="Item"
          onError={[Function]}
          onUpdate={[Function]}
          required={false}
          scrollIntoView={false}
        >
          <Field
            adjustFor="labels"
            className=""
            comments={false}
            commentsActive={false}
            commentsAdd="comments.add"
            commentsName="Comments"
            commentsRemove="comments.remove"
            commentsRequired={false}
            commentsValue={Object {}}
            errorPrefix=""
            errors={Array []}
            filterErrors={[Function]}
            help=""
            helpActive={false}
            optional={false}
            optionalText=""
            scrollIntoView={true}
            shrink={false}
            title="Provide position title"
            titleSize="h4"
            validate={true}
          >
            <Text
              bind={true}
              className="text"
              label="Position"
              name="Position"
              onError={[Function]}
              onUpdate={[Function]}
              prefilter={[Function]}
              prefix=""
              required={false}
              status={true}
              value=""
            />
          </Field>
          <Field
            adjustFor="labels"
            className=""
            comments={false}
            commentsActive={false}
            commentsAdd="comments.add"
            commentsName="Comments"
            commentsRemove="comments.remove"
            commentsRequired={false}
            commentsValue={Object {}}
            errorPrefix=""
            errors={Array []}
            filterErrors={[Function]}
            help=""
            helpActive={false}
            optional={false}
            optionalText=""
            scrollIntoView={true}
            shrink={false}
            title="Provide supervisor"
            titleSize="h4"
            validate={true}
          >
            <Text
              bind={true}
              className="text"
              label="Supervisor"
              name="Supervisor"
              onError={[Function]}
              onUpdate={[Function]}
              prefilter={[Function]}
              prefix=""
              required={false}
              status={true}
              value=""
            />
          </Field>
          <Field
            adjustFor="daterange"
            className=""
            comments={false}
            commentsActive={false}
            commentsAdd="comments.add"
            commentsName="Comments"
            commentsRemove="comments.remove"
            commentsRequired={false}
            commentsValue={Object {}}
            errorPrefix=""
            errors={Array []}
            filterErrors={[Function]}
            help="history.employment.default.additionalActivity.datesEmployed.help"
            helpActive={false}
            optional={false}
            optionalText=""
            scrollIntoView={true}
            shrink={true}
            title="Provide dates of employment"
            titleSize="h4"
            validate={true}
          >
            <DateRange
              allowPresent={false}
              bind={true}
              dateRangePrefix="additionalActivity"
              disabled={false}
              from={
                Object {
                  "touched": false,
                }
              }
              maxDate={
                Object {
                  "day": "15",
                  "month": "4",
                  "year": "2019",
                }
              }
              maxDateEqualTo={true}
              minDate={null}
              minDateEqualTo={true}
              name="DatesEmployed"
              onError={[Function]}
              prefix=""
              present={false}
              relationship=""
              to={
                Object {
                  "touched": false,
                }
              }
            />
          </Field>
        </AccordionItem>
      </BranchCollection>
    </div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <AdditionalActivity
        List={Object {}}
        employmentFromDate={
          Object {
            "day": "15",
            "month": "4",
            "year": "2019",
          }
        }
        onError={[Function]}
        onUpdate={[Function]}
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": AdditionalActivity {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "handleBlur": [Function],
        "handleChange": [Function],
        "handleFocus": [Function],
        "handleKeyDown": [Function],
        "handleValidation": [Function],
        "props": Object {
          "List": Object {},
          "employmentFromDate": Object {
            "day": "15",
            "month": "4",
            "year": "2019",
          },
          "onError": [Function],
          "onUpdate": [Function],
        },
        "refs": Object {},
        "state": null,
        "updateBranch": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="activity"
        >
          <BranchCollection
            List={Object {}}
            appendContent={null}
            appendLabel="Add additional periods"
            appendSize="h4"
            branchClassName=""
            branchName="branchcollection"
            content={null}
            employmentFromDate={
              Object {
                "day": "15",
                "month": "4",
                "year": "2019",
              }
            }
            help=""
            items={Array []}
            label="Do you have any additional periods of activity to add?"
            labelSize="h4"
            onError={[Function]}
            onUpdate={[Function]}
            removable={true}
            scrollToBottom=""
            valueKey="Has"
          >
            <AccordionItem
              bind={true}
              defaultState={true}
              dispatch={[Function]}
              name="Item"
              onError={[Function]}
              onUpdate={[Function]}
              required={false}
              scrollIntoView={false}
            >
              <Field
                adjustFor="labels"
                className=""
                comments={false}
                commentsActive={false}
                commentsAdd="comments.add"
                commentsName="Comments"
                commentsRemove="comments.remove"
                commentsRequired={false}
                commentsValue={Object {}}
                errorPrefix=""
                errors={Array []}
                filterErrors={[Function]}
                help=""
                helpActive={false}
                optional={false}
                optionalText=""
                scrollIntoView={true}
                shrink={false}
                title="Provide position title"
                titleSize="h4"
                validate={true}
              >
                <Text
                  bind={true}
                  className="text"
                  label="Position"
                  name="Position"
                  onError={[Function]}
                  onUpdate={[Function]}
                  prefilter={[Function]}
                  prefix=""
                  required={false}
                  status={true}
                  value=""
                />
              </Field>
              <Field
                adjustFor="labels"
                className=""
                comments={false}
                commentsActive={false}
                commentsAdd="comments.add"
                commentsName="Comments"
                commentsRemove="comments.remove"
                commentsRequired={false}
                commentsValue={Object {}}
                errorPrefix=""
                errors={Array []}
                filterErrors={[Function]}
                help=""
                helpActive={false}
                optional={false}
                optionalText=""
                scrollIntoView={true}
                shrink={false}
                title="Provide supervisor"
                titleSize="h4"
                validate={true}
              >
                <Text
                  bind={true}
                  className="text"
                  label="Supervisor"
                  name="Supervisor"
                  onError={[Function]}
                  onUpdate={[Function]}
                  prefilter={[Function]}
                  prefix=""
                  required={false}
                  status={true}
                  value=""
                />
              </Field>
              <Field
                adjustFor="daterange"
                className=""
                comments={false}
                commentsActive={false}
                commentsAdd="comments.add"
                commentsName="Comments"
                commentsRemove="comments.remove"
                commentsRequired={false}
                commentsValue={Object {}}
                errorPrefix=""
                errors={Array []}
                filterErrors={[Function]}
                help="history.employment.default.additionalActivity.datesEmployed.help"
                helpActive={false}
                optional={false}
                optionalText=""
                scrollIntoView={true}
                shrink={true}
                title="Provide dates of employment"
                titleSize="h4"
                validate={true}
              >
                <DateRange
                  allowPresent={false}
                  bind={true}
                  dateRangePrefix="additionalActivity"
                  disabled={false}
                  from={
                    Object {
                      "touched": false,
                    }
                  }
                  maxDate={
                    Object {
                      "day": "15",
                      "month": "4",
                      "year": "2019",
                    }
                  }
                  maxDateEqualTo={true}
                  minDate={null}
                  minDateEqualTo={true}
                  name="DatesEmployed"
                  onError={[Function]}
                  prefix=""
                  present={false}
                  relationship=""
                  to={
                    Object {
                      "touched": false,
                    }
                  }
                />
              </Field>
            </AccordionItem>
          </BranchCollection>
        </div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="activity"
        >
          <BranchCollection
            List={Object {}}
            appendContent={null}
            appendLabel="Add additional periods"
            appendSize="h4"
            branchClassName=""
            branchName="branchcollection"
            content={null}
            employmentFromDate={
              Object {
                "day": "15",
                "month": "4",
                "year": "2019",
              }
            }
            help=""
            items={Array []}
            label="Do you have any additional periods of activity to add?"
            labelSize="h4"
            onError={[Function]}
            onUpdate={[Function]}
            removable={true}
            scrollToBottom=""
            valueKey="Has"
          >
            <AccordionItem
              bind={true}
              defaultState={true}
              dispatch={[Function]}
              name="Item"
              onError={[Function]}
              onUpdate={[Function]}
              required={false}
              scrollIntoView={false}
            >
              <Field
                adjustFor="labels"
                className=""
                comments={false}
                commentsActive={false}
                commentsAdd="comments.add"
                commentsName="Comments"
                commentsRemove="comments.remove"
                commentsRequired={false}
                commentsValue={Object {}}
                errorPrefix=""
                errors={Array []}
                filterErrors={[Function]}
                help=""
                helpActive={false}
                optional={false}
                optionalText=""
                scrollIntoView={true}
                shrink={false}
                title="Provide position title"
                titleSize="h4"
                validate={true}
              >
                <Text
                  bind={true}
                  className="text"
                  label="Position"
                  name="Position"
                  onError={[Function]}
                  onUpdate={[Function]}
                  prefilter={[Function]}
                  prefix=""
                  required={false}
                  status={true}
                  value=""
                />
              </Field>
              <Field
                adjustFor="labels"
                className=""
                comments={false}
                commentsActive={false}
                commentsAdd="comments.add"
                commentsName="Comments"
                commentsRemove="comments.remove"
                commentsRequired={false}
                commentsValue={Object {}}
                errorPrefix=""
                errors={Array []}
                filterErrors={[Function]}
                help=""
                helpActive={false}
                optional={false}
                optionalText=""
                scrollIntoView={true}
                shrink={false}
                title="Provide supervisor"
                titleSize="h4"
                validate={true}
              >
                <Text
                  bind={true}
                  className="text"
                  label="Supervisor"
                  name="Supervisor"
                  onError={[Function]}
                  onUpdate={[Function]}
                  prefilter={[Function]}
                  prefix=""
                  required={false}
                  status={true}
                  value=""
                />
              </Field>
              <Field
                adjustFor="daterange"
                className=""
                comments={false}
                commentsActive={false}
                commentsAdd="comments.add"
                commentsName="Comments"
                commentsRemove="comments.remove"
                commentsRequired={false}
                commentsValue={Object {}}
                errorPrefix=""
                errors={Array []}
                filterErrors={[Function]}
                help="history.employment.default.additionalActivity.datesEmployed.help"
                helpActive={false}
                optional={false}
                optionalText=""
                scrollIntoView={true}
                shrink={true}
                title="Provide dates of employment"
                titleSize="h4"
                validate={true}
              >
                <DateRange
                  allowPresent={false}
                  bind={true}
                  dateRangePrefix="additionalActivity"
                  disabled={false}
                  from={
                    Object {
                      "touched": false,
                    }
                  }
                  maxDate={
                    Object {
                      "day": "15",
                      "month": "4",
                      "year": "2019",
                    }
                  }
                  maxDateEqualTo={true}
                  minDate={null}
                  minDateEqualTo={true}
                  name="DatesEmployed"
                  onError={[Function]}
                  prefix=""
                  present={false}
                  relationship=""
                  to={
                    Object {
                      "touched": false,
                    }
                  }
                />
              </Field>
            </AccordionItem>
          </BranchCollection>
        </div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <AdditionalActivity
    List={Object {}}
    employmentFromDate={
      Object {
        "day": "15",
        "month": "4",
        "year": "2019",
      }
    }
    onError={[Function]}
    onUpdate={[Function]}
  />,
}
`;
